File: output_20260318_123118/claude-opus/c/Linklist/map.c
#include "stdlib.h"
#include "assert.h"

struct node {
    struct node *next;
    int value;
};

/*@ 
predicate nodes(struct node *l; list<int> vs) =
    l == 0 ? vs == nil :
    l->value |-> ?v &*& l->next |-> ?n &*& malloc_block_node(l) &*& nodes(n, ?vs0) &*& vs == cons(v, vs0);
@*/

struct node *list_cons(int value, struct node *next)
    //@ requires nodes(next, ?vs);
    //@ ensures nodes(result, cons(value, vs));
{
    struct node *result = (struct node *)malloc(sizeof(struct node)); 
    if (result == 0) { abort(); }
    result->value = value;
    result->next = next;
    
    return result;
}

bool equals(struct node *n1, struct node *n2)
    //@ requires nodes(n1, ?vs1) &*& nodes(n2, ?vs2);
    //@ ensures nodes(n1, vs1) &*& nodes(n2, vs2) &*& result == (vs1 == vs2);
{
    bool result = false;
    if (n1 == 0)
        result = n2 == 0;
    else if (n2 == 0)
        result = false;
    else if (n1->value != n2->value)
        result = false;
    else {
        bool tmp = equals(n1->next, n2->next);
        result = tmp;
    }
    return result;
}

void dispose(struct node *l)
    //@ requires nodes(l, ?vs);
    //@ ensures true;
{
    if (l != 0) {
        struct node *next = l->next;
        //@ open nodes(l, ?vs0);
        free(l);
        dispose(next);
    }
}

typedef int (* mapfunc)(void *data, int x);

/*@ 
fixpoint list<int> map_list(list<int> vs, int f(void *data, int x), void *data) {
    switch(vs) {
        case nil: return nil;
        case cons(h,t): return cons(f(data,h), map_list(t,f,data));
    }
}
@*/

struct node *fmap(struct node *list, mapfunc f, void *data)
    //@ requires nodes(list, ?vs) &*& f(data, 0) == f(data, 0);
    //@ ensures nodes(list, vs) &*& nodes(result, map_list(vs, f, data));
{
    if (list == 0) {
        return 0;
    } else {
        int fvalue = f(data, list->value);
        struct node *fnext = fmap(list->next, f, data);
        struct node *result = list_cons(fvalue, fnext);
        return result;
    }
}

int plusOneFunc(void *data, int x) 
    //@ requires true;
    //@ ensures x == INT_MAX ==> false;
    //@ ensures result == x + 1;
{
    if (x == INT_MAX) abort();
    return x + 1;
}

int main() 
    //@ requires true;
    //@ ensures true;
{
    struct node *l = 0;
    //@ close nodes(l, nil);
    l = list_cons(3, l);
    l = list_cons(2, l);
    l = list_cons(1, l);
    
    struct node *l2 = fmap(l, plusOneFunc, 0);
    struct node *l3 = 0;
    //@ close nodes(l3, nil);
    l3 = list_cons(4, l3);
    l3 = list_cons(3, l3);
    l3 = list_cons(2, l3);
    
    bool tmp = equals(l2, l3);
    assert(tmp);
    
    dispose(l);
    dispose(l2);
    dispose(l3);
    return 0;
}

File: output_20260318_123118/claude-opus/rust/Stack/map.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Node {
    next: *mut Node,
    value: i32,
}

predicate nodes(struct Node *n; list<int> vs) =
    n == std::ptr::null_mut()
    ?
        vs == []
    :
        n |-> Node{next: *mut Node, value: int} &*&
        nodes((*n).next, ?rest) &*&
        vs == cons((*n).value, rest);

struct Stack {
    head: *mut Node,
}

predicate stack(struct Stack *s; list<int> vs) =
    s |-> Stack{head: *mut Node} &*& nodes((*s).head, vs);

type I32Func = unsafe fn(*mut u8, i32) -> i32;

unsafe fn map_nodes(n: *mut Node, f: I32Func, data: *mut u8)
    #[requires(nodes(n, ?vs))]
    #[ensures(nodes(n, ?vs2))]
    //@ decreases length(vs)
{
    if !n.is_null() {
        int old_val = (*n).value;
        int y = f(data, old_val);
        (*n).value = y;
        map_nodes((*n).next, f, data);
        //@ assert nodes((*n).next, ?rest);
        //@ close nodes(n, cons(y, rest));
    }
    else {
        //@ close nodes(n, []);
    }
}

unsafe fn dispose_nodes(n: *mut Node)
    #[requires(nodes(n, _))]
    #[ensures(true)]
    //@ decreases nodes_length(n)
{
    if !n.is_null() {
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {

    unsafe fn create() -> *mut Stack
        #[requires(true)]
        #[ensures(stack(result, []))]
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        //@ close nodes((*stack).head, []);
        //@ close stack(stack, []);
        return stack;
    }
    
    unsafe fn push(stack: *mut Stack, value: i32)
        #[requires(stack(stack, ?vs))]
        #[ensures(stack(stack, cons(value, vs)))]
    {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
        //@ close nodes(n, cons(value, vs));
        //@ close stack(stack, cons(value, vs));
    }

    unsafe fn pop(stack: *mut Stack) -> i32
        #[requires(stack(stack, ?vs)) &*& vs != []]
        #[ensures(stack(stack, tail(vs)) &*& result == head(vs))]
    {
        let head = (*stack).head;
        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());
        //@ open nodes(head, vs);
        //@ open stack(stack, vs);
        //@ close stack(stack, tail(vs));
        return result;
    }
    
    unsafe fn map(stack: *mut Stack, f: I32Func, data: *mut u8)
        #[requires(stack(stack, ?vs))]
        #[ensures(stack(stack, ?vs2))]
        //@ decreases length(vs)
    {
        map_nodes((*stack).head, f, data);
        //@ open stack(stack, vs);
        //@ close stack(stack, vs);
    }
    
    unsafe fn dispose(stack: *mut Stack)
        #[requires(stack(stack, _))]
        #[ensures(true)]
    {
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }

}

unsafe fn plus_a(data: *mut u8, x: i32) -> i32
    #[requires(true)]
    #[ensures(true)]
{
    let result = x + *(data as *mut i32);
    result
}

unsafe fn read_i32() -> i32
    #[requires(true)]
    #[ensures(true)]
{
    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    line.parse().unwrap()
}

fn main() {
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        Stack::push(s, 30);
        let mut a = read_i32();

        Stack::map(s, plus_a, &raw mut a as *mut u8);

        Stack::dispose(s);
    }
}

fixpoint int head(list<int> vs) {
    switch(vs) {
        case nil: return 0; 
        case cons(x, _): return x;
    }
}

fixpoint list<int> tail(list<int> vs) {
    switch(vs) {
        case nil: return nil;
        case cons(_, xs): return xs;
    }
}

fixpoint int nodes_length(struct Node *n) {
    return n == std::ptr::null_mut() ? 0 : 1 + nodes_length((*n).next);
}